pallets/api/src/extension.rs

```diff
@@ -136,17 +138,187 @@ fn version(env: &impl Environment) -> u8 {
+	}
+
 	#[test]
 	fn prepender_works() {
 		let env = MockEnvironment {
-			func_id: u16::from_le_bytes([0, 1]),
-			ext_id: u16::from_le_bytes([2, 3]),
+			func_id: u16::from_le_bytes([1, 2]),
+			ext_id: u16::from_le_bytes([3, 4]),
 		};
-		assert_eq!(Prepender::process(vec![0u8], &env), vec![1, 2, 3, 0]);
+		assert_eq!(Prepender::process(vec![0u8], &env), vec![2, 3, 4, 0]);
+		assert_eq!(Prepender::process(vec![0u8, 5, 10], &env), vec![2, 3, 4, 0, 5, 10]);
+	}
+
+	#[test]
+	fn identified_by_first_byte_of_function_id_matches() {
+		let env = MockEnvironment { func_id: u16::from_le_bytes([1, 2]), ext_id: 0u16 };
+		assert!(IdentifiedByFirstByteOfFunctionId::<ConstU8<1>>::matches(&env));
+	}
+
+	#[test]
+	fn identified_by_first_byte_of_function_id_does_not_match() {
+		let env = MockEnvironment { func_id: u16::from_le_bytes([1, 2]), ext_id: 0u16 };
+		assert!(!IdentifiedByFirstByteOfFunctionId::<ConstU8<2>>::matches(&env));
+	}
+
+	#[test]
+	fn dispatch_call_log_target_works() {
+		assert!(matches!(
+			<DispatchCallLogTarget as LogTarget>::LOG_TARGET,
+			"pop-api::extension::dispatch"
+		));
+	}
+
+	#[test]
+	fn read_state_log_target_works() {
+		assert!(matches!(
+			<ReadStateLogTarget as LogTarget>::LOG_TARGET,
+			"pop-api::extension::read-state"
+		));
+	}
+
+	mod versioned_error {
+		use super::{RetVal::Converging, *};
+
+		// Mock versioned error.
+		#[derive(Debug)]
+		pub enum VersionedError {
+			V0(DispatchError),
+			V1(DispatchError),
+		}
+
+		impl TryFrom<(DispatchError, u8)> for VersionedError {
+			type Error = DispatchError;
+
+			fn try_from(value: (DispatchError, u8)) -> Result<Self> {
+				let (error, version) = value;
+				match version {
+					0 => Ok(VersionedError::V0(error)),
+					1 => Ok(VersionedError::V1(error)),
+					_ => Err(Other("DecodingFailed")),
+				}
+			}
+		}
+
+		impl From<VersionedError> for u32 {
+			// Mock conversion based on error and version.
+			fn from(value: VersionedError) -> Self {
+				match value {
+					VersionedError::V0(error) => match error {
+						BadOrigin => 1,
+						_ => 100,
+					},
+					VersionedError::V1(error) => match error {
+						BadOrigin => 2,
+						_ => 200,
+					},
+				}
+			}
+		}
+
+		#[test]
+		fn versioned_error_converter_works() {
+			for (version, error, expected_result) in vec![
+				(0, BadOrigin, 1),
+				(0, CannotLookup, 100),
+				(1, BadOrigin, 2),
+				(1, CannotLookup, 200),
+			] {
+				let env =
+					MockEnvironment { func_id: u16::from_le_bytes([0, version]), ext_id: 0u16 };
+				// Because `Retval` does not implement the `Debug` trait the result has to be
+				// unwrapped.
+				let Ok(Converging(result)) =
+					VersionedErrorConverter::<VersionedError>::convert(error, &env)
+				else {
+					panic!("should not happen")
+				};
+				assert_eq!(result, expected_result);
+			}
+		}
+
+		#[test]
+		fn versioned_error_converter_fails() {
+			let env = MockEnvironment { func_id: u16::from_le_bytes([0, 2]), ext_id: 0u16 };
```

Comment:
```suggestion
			let version = 2;
			let env = MockEnvironment { func_id: u16::from_le_bytes([0, version]), ext_id: 0u16 };
```
Makes it a bit easier to see what is going on.